src/code/padmgr.c

```diff
@@ -1,20 +1,81 @@
+/**
+ * @file padmgr.c
+ *
+ * This file implements communicating with joybus devices at a high level and serving the results to other threads.
+ *
+ * Any device that can be plugged into one of the four controller ports such as a standard N64 controller is a joybus
+ * device. Some joybus devices are also located inside the cartridge such as EEPROM for save data or the Real-Time
+ * Clock, however neither of these are used in Zelda64 and so this type of communication is unimplemented. Of the
+ * possible devices that can be plugged into the controller ports, the only device that padmgr will recognize and
+ * attempt to communicate with is the standard N64 controller.
+ *
+ * Communicating with these devices is broken down into various layers:
+ *
+ * Other threads                    : The rest of the program that will use the polled data
+ *  |
+ * PadMgr                           : Manages devices, submits polling commands at vertical retrace
+ *  |
+ * Libultra osCont* routines        : Interface for building commands and safely using the Serial Interface
+ *  |
+ * Serial Interface                 : Hardware unit for sending joybus commands and receiving data via DMA
+ *  |
+ * PIF                              : Forwards joybus commands and receives response data from the devices
+ *  |---¬---¬---¬-------¬
+ *  1   2   3   4       5           : The joybus devices plugged into the four controller ports or on the cartridge
+ *
+ * Joybus communication is handled on another thread as polling and receiving controller data is a slow process; the
+ * N64 programming manual section 26.2.4.1 quotes 2 milliseconds as the expected delay from calling
+ * `osContStartReadData` to receiving the data. By running this on a separate thread to the game state, work can be
+ * done while waiting for this operation to complete.
+ */
 #include "global.h"
 #include "vt.h"
 
-s32 D_8012D280 = 1;
-
-OSMesgQueue* PadMgr_LockSerialMesgQueue(PadMgr* padMgr) {
+#define PADMGR_LOG(controllerNo, msg)                                    \
+    if (1) {                                                             \
+        osSyncPrintf(VT_FGCOL(YELLOW));                                  \
+        /* padmgr: Controller %d: %s */                                  \
+        osSyncPrintf("padmgr: %dコン: %s\n", (controllerNo) + 1, (msg)); \
+        osSyncPrintf(VT_RST);                                            \
+    }(void)0
+
+#define LOG_SEVERITY_NOLOG 0
+#define LOG_SEVERITY_INFO 1
+#define LOG_SEVERITY_ERROR 2
+#define LOG_SEVERITY_VERBOSE 3
+
+s32 gPadMgrLogSeverity = LOG_SEVERITY_INFO;
```

Comment:
For similar reasons to above I'll hold off on this change, since all other `*LogSeverity` variables are currently `g`-prefixed. We can look into changing them all at once some other time.